src/vvwe.rs

```diff
@@ -107,24 +108,43 @@ impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
         match op.happens_after() {
             // Dang! we have a happens after relation!
             Some(dot) => {
-                // Let's buffer this operation then.
+                // Have we already seen the operation we depend on?
                 if !self.saw_site_dot(&dot) {
-                    self.buffer.insert(dot, op);
-                    // and do nothing
-                    None
+                    // No, so buffer this operation until we see it
+                    self.buffer.entry(dot)
+                        .or_default()
+                        .push(op);
+                    vec![]
                 } else {
-                    Some(op)
+                    // Yes! Free this operation and all other operations which depend on this one
+                    self.remove_from_buffer(op)
                 }
             }
             None => {
-                // Ok so we're not causally constrained, but maybe we already saw an associated
-                // causal operation? If so let's just delete the pair
-                match self.buffer.remove(&op.dot()) {
-                    Some(_) => None, // we are dropping the dependent op! that can't be right
-                    None => Some(op),
-                }
+                // The operation we're processing doesn't depend on anything but maybe we've
+                // already seen operations which depend on this one? If so return those and free
+                // things transitively.
+                self.remove_from_buffer(op)
+            }
+        }
+    }
+
+    //TODO: better name
+    fn remove_from_buffer(&mut self, op: T) -> Vec<T> {
+        let mut queue = VecDeque::new();
+        queue.push_back(op.dot());
+
+        let mut cleared_ops = vec![op];
+        // Search the buffer in bread-first order for operations we can now perform after freeing
+        // up dot
+        while let Some(root) = queue.pop_front() {
```

Comment:
I wrote this using a queue since I'm not sure how well rust would handle a recursive bfs